Mathlib/Tactic/Widget/StringDiagram.lean

```diff
@@ -0,0 +1,356 @@
+  for (l₁, l₂) in pairs nodes do
+    if let .some x₁ := l₁.head? then
+      if let .some x₂ := l₂.head? then
+        addInstruction s!"Above({x₁.toPenroseVar}, {x₂.toPenroseVar})"
+  /- Add 1-morphisms as strings. -/
+  for l in strands do
+    for s in l do
+      addConstructor "Mor1" s.toPenroseVar
+        "MakeString" [s.startPoint.toPenroseVar, s.endPoint.toPenroseVar]
+
+/-- Penrose dsl file for string diagrams. -/
+def dsl :=
+  include_str ".."/".."/".."/"widget"/"src"/"penrose"/"monoidal.dsl"
+
+/-- Penrose sty file for string diagrams. -/
+def sty :=
+  include_str ".."/".."/".."/"widget"/"src"/"penrose"/"monoidal.sty"
+
+open scoped Jsx in
+/-- Construct a string diagram from the expression of a 2-morphism. -/
+def fromExpr (e : Expr) : MonoidalM Html := do
+  let e' := (← eval e).expr
+  DiagramBuilderM.run do
+    mkStringDiagram e'
+    match ← DiagramBuilderM.buildDiagram dsl sty with
+    | some html => return html
+    | none => return <span>No non-structural morphisms found.</span>
+
+/-- Given a 2-morphism, return a string diagram. Otherwise `none`. -/
+def stringM? (e : Expr) : MetaM (Option Html) := do
+  let e ← instantiateMVars e
+  let some ctx ← mkContext? e | return none
+  return some <| ← MonoidalM.run ctx <| fromExpr e
+
+open scoped Jsx in
+/-- Help function for displaying two string diagrams in an equality. -/
+def mkEqHtml (lhs rhs : Html) : Html :=
+  <div className="flex">
+    <div className="w-50">
+      <details «open»={true}>
+        <summary className="mv2 pointer">String diagram for LHS</summary> {lhs}
+      </details>
+    </div>
+    <div className="w-50">
+      <details «open»={true}>
+        <summary className="mv2 pointer">String diagram for RHS</summary> {rhs}
+      </details>
+    </div>
+  </div>
+
+/-- Given an equality between 2-morphisms, return a string diagram of the LHS and RHS.
+Otherwise `none`. -/
+def stringEqM? (e : Expr) : MetaM (Option Html) := do
+  let e ← instantiateMVars e
+  let some (_, lhs, rhs) := e.eq? | return none
+  let some lhs ← stringM? lhs | return none
+  let some rhs ← stringM? rhs | return none
+  return some <| mkEqHtml lhs rhs
+
+/-- Given an 2-morphism or equality between 2-morphisms, return a string diagram.
+Otherwise `none`. -/
+def stringMorOrEqM? (e : Expr) : MetaM (Option Html) := do
+  if let some html ← stringM? e then
+    return some html
+  else if let some html ← stringEqM? e then
+    return some html
+  else
+    return none
+
+/-- The `Expr` presenter for displaying string diagrams. -/
+@[expr_presenter]
+def stringPresenter : ExprPresenter where
+  userName := "String diagram"
+  layoutKind := .block
+  present type := do
+    if let some html ← stringMorOrEqM? type then
+      return html
+    throwError "Couldn't find a 2-morphism to display a string diagram."
+
+open scoped Jsx in
+/-- The RPC method for displaying string diagrams. -/
+@[server_rpc_method]
+def rpc (props : PanelWidgetProps) : RequestM (RequestTask Html) :=
```

Comment:
Since `GoalTypePanel` can also draw this now using the presenter, should we get rid of this and the `StringDiagram` component?

Reply:
I see! Yes, there is no good way to debug `ExprPresenter`s right now. I am assuming that you need to keep the rpc method if you want to keep `StringDiagram`, as the rpc method is its main implementation!